shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -736,15 +731,21 @@ - (void)loadAOTData:(NSString*)assetsDir {
 - (void)registerViewController:(FlutterViewController*)controller
                  forIdentifier:(FlutterViewIdentifier)viewIdentifier {
   NSAssert(controller != nil, @"The controller must not be nil.");
-  NSAssert(![controller attached],
-           @"The incoming view controller is already attached to an engine.");
+  NSAssert(controller.engine == nil,
+           @"The FlutterViewController is unexpectedly attached to "
+           @"engine %@ before initialization.",
+           controller.engine);
   NSAssert([_viewControllers objectForKey:@(viewIdentifier)] == nil,
            @"The requested view ID is occupied.");
+  [_viewControllers setObject:controller forKey:@(viewIdentifier)];
   [controller setUpWithEngine:self
                viewIdentifier:viewIdentifier
            threadSynchronizer:_threadSynchronizer];
   NSAssert(controller.viewIdentifier == viewIdentifier, @"Failed to assign view ID.");
-  [_viewControllers setObject:controller forKey:@(viewIdentifier)];
+  NSAssert(controller.attached && controller.engine == self,
+           @"The FlutterViewController unexpectedly stays unattached after being added. "
```

Comment:
This assertion message makes sense when you're reading this code, but it might be a little confusing if you see this in the command line without any other context.

Perhaps consider expanding the message to explain this is attaching/adding to the engine. Consider something like:

> The FlutterViewController should switch to the attached mode after it is added to a FlutterEngine.

Same comment for the other similar assertions